FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,51 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { (context) in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { (context) in
+			do {
+				let cache = try ManagedCache.newUniqueInstance(in: context)
+				cache.feed = ManagedFeedImage.images(from: feed, in: context)
+				cache.timestamp = timestamp
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { (context) in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					context.delete(cache)
+				}
+				try context.save()
```

Comment:
`try context.save()` should be inside the if statement block because the context should only be saved if there are changes to it. Otherwise, Core Data may perform unnecessary work.

 ```suggestion
				if let cache = try ManagedCache.find(in: context) {
					context.delete(cache)
					try context.save()
				}
```

Reply:
Thank you guys for providing is much value 
I'm learning a lot